homeassistant/components/picnic/services.py

```diff
@@ -0,0 +1,131 @@
+import homeassistant.helpers.config_validation as cv
+
+from .const import (
+    ATTR_AMOUNT,
+    ATTR_DEVICE_ID,
+    ATTR_PRODUCT_ID,
+    ATTR_PRODUCT_IDENTIFIERS,
+    ATTR_PRODUCT_NAME,
+    CONF_API,
+    DOMAIN,
+    SERVICE_ADD_PRODUCT_TO_CART,
+)
+
+
+class PicnicServiceException(Exception):
+    """Exception for Picnic services."""
+
+
+async def async_register_services(hass: HomeAssistant) -> None:
+    """Register services for the Picnic integration, if not registered yet."""
+
+    # Register the add_product service
+    if not hass.services.has_service(DOMAIN, SERVICE_ADD_PRODUCT_TO_CART):
+
+        async def async_add_product_service(call: ServiceCall):
+            # Get the picnic API client and call the handler with api client and hass
+            api_client: PicnicAPI = await get_api_client(
+                hass, call.data.get("device_id")
+            )
+            await handle_add_product(hass, api_client, call)
+
+        hass.services.async_register(
+            DOMAIN,
+            SERVICE_ADD_PRODUCT_TO_CART,
+            async_add_product_service,
+            schema=vol.Schema(
+                {
+                    vol.Optional(ATTR_DEVICE_ID): cv.string,
+                    vol.Exclusive(
+                        ATTR_PRODUCT_ID, ATTR_PRODUCT_IDENTIFIERS
+                    ): cv.positive_int,
+                    vol.Exclusive(
+                        ATTR_PRODUCT_NAME, ATTR_PRODUCT_IDENTIFIERS
+                    ): cv.string,
+                    vol.Optional(ATTR_AMOUNT): vol.All(
+                        vol.Coerce(int), vol.Range(min=1)
+                    ),
+                }
+            ),
+        )
+
+
+async def get_api_client(hass: HomeAssistant, device_id: Any | None = None):
+    """Get the right Picnic API client based on the device id, else get the default one."""
+    if device_id is None:
+        default_config_id = list(hass.data[DOMAIN].keys())[0]
+        return hass.data[DOMAIN][default_config_id][CONF_API]
+
+    # Get device from registry
+    registry = device_registry.async_get(hass)
+    device = registry.async_get(device_id)
+
+    if device is None:
+        raise PicnicServiceException(f"Device with id {device_id} not found!")
+
+    # Get Picnic API client for the config entry id
+    try:
+        config_entry_id = next(iter(device.config_entries))
+        return hass.data[DOMAIN][config_entry_id][CONF_API]
+    except (AttributeError, StopIteration, KeyError) as error:
+        raise PicnicServiceException(
+            f"Device with id {device_id} not found!"
+        ) from error
+
+
+async def handle_add_product(
+    hass: HomeAssistant, api_client: PicnicAPI, call: ServiceCall
+) -> None:
+    """Handle the call for the add_product service."""
+    product_id = call.data.get("product_id")
+    if not product_id:
+        search_results = await hass.async_add_executor_job(
+            _product_search, api_client, call.data.get("product_name")
```

Comment:
Given that `_product_search` does more than is needed right now, simplify it to just return the first product id in the result and remove the extra unused functionality, and the next few lines of code pulling out the product id can be removed here too, as well as the str() call below.

Reply:
I've simplified the method, but the `str()` call is still needed as the product id passed to the service can be an `int`.